monai/losses/image_dissimilarity.py

```diff
@@ -0,0 +1,153 @@
+        sigma = self.kernel_size / 3
+
+        grid_ndim = torch.arange(0, self.kernel_size)
+        grid_ndim_ch = torch.arange(0, self.in_channels)
+
+        if self.ndim == 1:
+            grid = torch.meshgrid(grid_ndim_ch, grid_ndim)
+        elif self.ndim == 2:
+            grid = torch.meshgrid(grid_ndim_ch, grid_ndim, grid_ndim)
+        elif self.ndim == 3:
+            grid = torch.meshgrid(grid_ndim_ch, grid_ndim, grid_ndim, grid_ndim)
+        else:
+            raise ValueError
+
+        grid = torch.stack(grid, dim=-1).to(dtype=torch.float)
+        kernel = torch.exp(-torch.sum(torch.square(grid - mean), dim=-1) / (2 * sigma ** 2)).unsqueeze(
+            0
+        )  # (1, in_channel, kernel_size, kernel_size, kernel_size)
+        return kernel, torch.sum(kernel ** 2), int((self.kernel_size - 1) / 2)
+
+    def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            input: the shape should be BNH[WD].
+            target: the shape should be BNH[WD].
+        Raises:
+            ValueError: When ``self.reduction`` is not one of ["mean", "sum", "none"].
+        """
+        assert (
+            input.shape[1] == self.in_channels
+        ), f"expecting input with {self.in_channels} channels, got input of shape {input.shape}"
+        assert (
+            input.ndim - 2 == self.ndim
+        ), f"expecting input with {self.ndim} spatial dimensions, got input of shape {input.shape}"
+        assert (
+            target.shape == input.shape
+        ), f"ground truth has differing shape ({target.shape}) from input ({input.shape})"
+
+        t2, p2, tp = target ** 2, input ** 2, target * input
+
+        # sum over kernel
+        fn = conv_dict[self.ndim]
+        t_sum = fn(target, weight=self.kernel, padding=self.padding)
+        p_sum = fn(input, weight=self.kernel, padding=self.padding)
+        t2_sum = fn(t2, weight=self.kernel, padding=self.padding)
+        p2_sum = fn(p2, weight=self.kernel, padding=self.padding)
+        tp_sum = fn(tp, weight=self.kernel, padding=self.padding)
+
+        # average over kernel
+        t_avg = t_sum / self.kernel_vol
+        p_avg = p_sum / self.kernel_vol
+
+        # normalized cross correlation between t and p
+        # sum[(t - mean[t]) * (p - mean[p])] / std[t] / std[p]
+        # denoted by num / denom
+        # assume we sum over N values
+        # num = sum[t * p - mean[t] * p - t * mean[p] + mean[t] * mean[p]]
+        #     = sum[t*p] - sum[t] * sum[p] / N * 2 + sum[t] * sum[p] / N
+        #     = sum[t*p] - sum[t] * sum[p] / N
+        #     = sum[t*p] - sum[t] * mean[p] = cross
+        # the following is actually squared ncc
+        cross = tp_sum - p_avg * t_sum
+        t_var = t2_sum - t_avg * t_sum  # std[t] ** 2
+        p_var = p2_sum - p_avg * p_sum  # std[p] ** 2
+        ncc = (cross * cross + EPS) / (t_var * p_var + EPS)  # shape = (batch, 1, D, H, W)
```

Comment:
perhaps make this EPS tunable, `self.eps` and defaulting to `1e-7`, e.g. https://github.com/Project-MONAI/MONAI/blob/db7bf8984712f3e46ed533611a70c9b8cd0293bd/monai/losses/dice.py#L51-L52

Reply:
updated